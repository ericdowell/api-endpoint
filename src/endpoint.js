import axios from 'axios'
import MessageBag from './errors/messageBag'
import qs from 'qs'

const trimSlashes = string =>
  (typeof string === 'string' && string.replace(/^\/|\/$/g, '')) || ''

const trimStartSlash = string =>
  (typeof string === 'string' && string.replace(/^\/+/g, '')) || ''

class Endpoint {
  /**
   *
   * @type {boolean}
   * @private
   */
  _debug = false

  /**
   *
   * @type {object}
   * @protected
   */
  _headers = {}

  /**
   *
   * @returns {this}
   */
  debug() {
    this._debug = true
    return this
  }

  /**
   *
   * @returns {boolean}
   */
  isDebugEnabled() {
    return this._debug === true
  }

  /**
   *
   * @param {object} headers
   * @returns {Endpoint}
   */
  setHeaders(headers) {
    this._headers = headers
    return this
  }

  /**
   *
   * @returns {object}
   */
  get headers() {
    return this._headers
  }

  /**
   *
   * @returns {boolean}
   */
  hasHeaders() {
    return Object.keys(this.headers).length < 0
  }

  /**
   * This contains the protocol and domain, aka location.origin.
   * e.g. https://example.com or https://localhost:3000
   *
   * @returns {string}
   */
  get origin() {
    return window.location.origin
  }

  /**
   * This it the middle of the url path, usually a group prefix.
   * e.g. api/v1 or user/settings
   *
   * @returns {string}
   */
  get path() {
    return ''
  }

  /**
   * This should be singular.
   * e.g. "user" or "profile" or this.constructor.name.toLowerCase()
   *
   * @returns {string}
   */
  get endpoint() {
    return ''
  }

  /**
   *
   * @returns {{AxiosStatic: AxiosStatic}}
   */
  get client() {
    return axios || window.axios
  }

  /**
   *
   * @returns {Console}
   */
  get output() {
    return window.console
  }

  /**
   * Used with axios library to set baseURL property.
   *
   * @returns {string}
   */
  getBaseUrl() {
    const origin = trimSlashes(this.origin)
    const path = trimSlashes(this.path)
    const endpoint = trimStartSlash(this.endpoint)
    if (!endpoint) {
      return `${origin}/${path}`
    }
    return `${origin}/${path}/${endpoint}`
  }

  /**
   *
   * @param {string} url
   * @param {string} method
   * @param {object=} options
   * @returns {object}
   */
  queryOptions(url, method, options) {
    options = options || {}
    options = Object.assign({}, options, { url, method })
    options.baseURL = this.getBaseUrl()
    if (!options.paramsSerializer) {
      options.paramsSerializer = params => {
        return qs.stringify(params, { arrayFormat: 'brackets' })
      }
    }
    if (this.hasHeaders()) {
      options.headers = Object.assign({}, options.headers || {}, this.headers)
    }
    if (!options.data || typeof options.data !== 'object') {
      return options
    }
    return options
  }

  /**
   *
   * @param {string} url
   * @param {string} method
   * @param {object=} options
   * @returns {Promise<any>}
   */
  query(url, method, options) {
    const client = this.client
    options = this.queryOptions(url, method, options)

    return client(options)
      .then(response => {
        if (this.isDebugEnabled()) {
          this.debugResponse(url, method, options, response)
        }
        return this.responseData(response)
      })
      .catch(error => {
        const responseError = this.responseError(url, method, options, error)
        if (this.isDebugEnabled()) {
          this.debugResponseError(url, method, options, responseError)
        }
        return Endpoint.handleQueryError(responseError, this)
      })
  }

  /**
   *
   * @param {object} response
   * @returns {*}
   */
  responseData(response) {
    return response && response.data
  }

  /**
   *
   * @param {string} url
   * @param {string} method
   * @param {object=} options
   * @param {AxiosError|Error} error
   * @returns {MessageBag}
   */
  responseError(url, method, options, error) {
    const messageBag = this.newMessageBag()
    messageBag.url = url
    messageBag.method = method
    messageBag.options = options
    error = error || { config: {}, response: {} }
    messageBag.original = error
    messageBag.config = (error && error.config) || {}
    messageBag.response = (error && error.response) || {}
    return messageBag
  }

  /**
   *
   * @returns {MessageBag}
   */
  newMessageBag() {
    return new MessageBag()
  }

  /**
   *
   * @param {string} url
   * @param {string} method
   * @param {object=} options
   * @param {object} response
   */
  debugResponse(url, method, options, response) {
    this.debugQuery(url, method, options, 'response', response)
  }

  /**
   *
   * @param {string} url
   * @param {string} method
   * @param {object=} options
   * @param {MessageBag} error
   */
  debugResponseError(url, method, options, error) {
    this.debugQuery(url, method, options, 'error', error)
  }

  /**
   *
   * @param {string} url
   * @param {string} method
   * @param {object=} options
   * @param {string} label
   * @param {object} data
   */
  debugQuery(url, method, options, label, data) {
    let output = [
      'color: #e55ea2;',
      url,
      'color: black;',
      'color: #e55ea2;',
      method,
      'color: black;',
      options
    ]
    let format = 'url: %c"%s"%c\nmethod: %c"%s"%c\noptions:\n%o'
    if (label && data) {
      format += '\n%s:\n%o'
      output.push(label)
      output.push(data)
    }
    this.output.log(format, ...output)
  }

  /**
   *
   * @param {Error} error
   * @param {object|Endpoint} object
   * @returns {void|*}
   * @throw {Error}
   */
  static handleQueryError(error, object) {
    if (
      !(object instanceof Endpoint) ||
      !object.handleError ||
      typeof object.handleError !== 'function'
    ) {
      throw error
    }
    return object.handleError(error)
  }
}

export default Endpoint
