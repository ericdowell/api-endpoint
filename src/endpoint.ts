import axios, {
  AxiosError,
  AxiosRequestConfig,
  AxiosResponse,
  Method,
} from 'axios'
import qs from 'qs'
import urljoin from 'url-join'

export class Endpoint {
  /**
   *
   * @type {boolean}
   * @private
   */
  _debug = false

  /**
   *
   * @type {object}
   * @protected
   */
  _headers: { [key: string]: any } = {}

  /**
   * @type {boolean}
   * @protected
   */
  _axiosResponse = false

  /**
   *
   * @returns {this}
   */
  axiosResponse(): this {
    this._axiosResponse = true
    return this
  }

  /**
   *
   * @returns {this}
   */
  enableDebug(): this {
    this._debug = true
    return this
  }

  /**
   *
   * @returns {boolean}
   */
  isDebugEnabled(): boolean {
    return this._debug
  }

  /**
   *
   * @param {object} headers
   * @returns {this}
   */
  setHeaders(headers: any): this {
    this._headers = headers
    return this
  }

  /**
   *
   * @param {string} key
   * @param {*} value
   * @returns {this}
   */
  setHeader(key: string, value: any): this {
    this._headers[key] = value
    return this
  }

  /**
   *
   * @returns {object}
   */
  get headers(): { [key: string]: any } {
    return this._headers
  }

  /**
   *
   * @returns {boolean}
   */
  hasHeaders(): boolean {
    return Object.keys(this.headers).length !== 0
  }

  /**
   * This contains the protocol and domain, aka location.origin.
   * e.g. https://example.com or https://localhost:3000
   *
   * @returns {string}
   */
  get origin(): string {
    return window.location.origin
  }

  /**
   * This is the middle of the url path, usually a group prefix.
   * e.g. api/v1 or user/settings
   *
   * @returns {string}
   */
  get path(): string {
    return ''
  }

  /**
   * This is the last part of the url path, usually the resource name.
   * e.g. "user" or "profile"
   *
   * @returns {string}
   */
  get endpoint(): string {
    return ''
  }

  /**
   * Returns console for output.
   *
   * @returns {Console}
   */
  get console(): Console {
    return (window && window.console) || console
  }

  /**
   * Used with axios library to set baseURL property.
   *
   * @returns {string}
   */
  get baseURL(): string {
    return urljoin(this.origin, this.path, this.endpoint)
  }

  /**
   * Returns object of query options, currently defaults to Axios options.
   *
   * @param {string} url
   * @param {string} method
   * @param {object=} requestConfig
   * @returns {object}
   */
  queryOptions(
    url: string,
    method: Method,
    requestConfig?: AxiosRequestConfig,
  ): AxiosRequestConfig {
    const config = { ...(requestConfig ?? {}), ...{ url, method } }
    config.baseURL = this.baseURL
    if (!config.paramsSerializer) {
      config.paramsSerializer = (params): string => {
        return qs.stringify(params, { arrayFormat: 'brackets' })
      }
    }
    if (this.hasHeaders()) {
      config.headers = { ...config.headers, ...this.headers }
    }
    return config
  }

  /**
   * General query method that is used by all HTTP calls.
   *
   * @param {string} url
   * @param {string} method
   * @param {object=} requestConfig
   * @returns {Promise<any>}
   */
  async query<T = any, R = AxiosResponse<T>>(
    url: string,
    method: Method,
    requestConfig?: AxiosRequestConfig | any,
  ): Promise<T | R> {
    const config = this.queryOptions(url, method, requestConfig)
    try {
      const response = await axios.request<T, R>(config)
      if (this.isDebugEnabled()) {
        this.debugResponse(url, method, config, response)
      }
      return this._axiosResponse ? response : this.responseData<T>(response)
    } catch (error) {
      if (this.isDebugEnabled()) {
        this.debugResponseError(url, method, config, error)
      }
      return Endpoint.handleQueryError(error, this)
    }
  }

  /**
   *
   * @param {object} response
   * @returns {any}
   */
  responseData<T = any>(response: AxiosResponse | any): T {
    return response?.data
  }

  /**
   *
   * @param {string} url
   * @param {string} method
   * @param {object=} config
   * @param {object} response
   */
  debugResponse<T = any, R = AxiosResponse<T>>(
    url: string,
    method: string,
    config: AxiosRequestConfig,
    response: R,
  ): void {
    this.debugQuery<R>(url, method, config, 'response', response)
  }

  /**
   *
   * @param {string} url
   * @param {string} method
   * @param {object=} config
   * @param {AxiosError} error
   */
  debugResponseError<T = AxiosError>(
    url: string,
    method: string,
    config: AxiosRequestConfig,
    error: T,
  ): void {
    this.debugQuery<T>(url, method, config, 'error', error)
  }

  /**
   *
   * @param {string} url
   * @param {string} method
   * @param {object=} config
   * @param {string} label
   * @param {object} data
   */
  debugQuery<T = any>(
    url: string,
    method: string,
    config: AxiosRequestConfig,
    label: string,
    data: T,
  ): void {
    const output = [
      'color: #e55ea2;',
      url,
      'color: black;',
      'color: #e55ea2;',
      method,
      'color: black;',
      config,
    ]
    let format = 'url: %c"%s"%c\nmethod: %c"%s"%c\nconfig:\n%o'
    if (label && data) {
      format += '\n%s:\n%o'
      output.push(label)
      output.push(data)
    }
    this.console.log(format, ...output)
  }

  handleError(error: AxiosError | Error): never {
    throw error
  }

  /**
   *
   * @param {Error} error
   * @param {any} endpointInstance
   * @returns {void|*}
   * @throw {Error}
   */
  static handleQueryError(
    error: AxiosError | Error,
    endpointInstance: Endpoint,
  ): any {
    return endpointInstance.handleError(error)
  }
}
