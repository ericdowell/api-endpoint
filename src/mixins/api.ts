import { AxiosRequestConfig } from 'axios'
import qs from 'qs'
import urljoin from 'url-join'
import { Constructor } from './types'

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function ApiMixin<T extends Constructor<any>>(superClass: T) {
  return class extends superClass {
    /** @protected */
    shouldStringify = true

    // Always set Content-Type to application/x-www-form-urlencoded, place in assign last.
    get config(): AxiosRequestConfig {
      return {
        ...super.config,
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded',
          ...super.config.headers,
          Accept: 'application/json',
        },
      }
    }

    // Override to set as version, default to empty
    // e.g. v1, v3, v5 and so on.
    get version(): string {
      return ''
    }

    // This it the middle of the url path, usually a group prefix.
    // e.g. api/v1 or user/settings
    get path(): string {
      return urljoin('api', this.version)
    }

    preventStringify(): this {
      this.shouldStringify = false
      return this
    }

    requestConfig(requestConfig: AxiosRequestConfig): AxiosRequestConfig {
      const config: AxiosRequestConfig = super.requestConfig(requestConfig)
      if (!config.data || typeof config.data !== 'object') {
        return config
      }
      if (this.shouldStringify && !(config.data instanceof FormData)) {
        config.data = qs.stringify(config.data)
      }
      return config
    }
  }
}
