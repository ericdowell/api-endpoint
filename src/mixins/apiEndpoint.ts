import { AxiosRequestConfig, Method } from 'axios'
import qs from 'qs'
import { Constructor } from './index'
import urljoin from 'url-join'

// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
export function ApiEndpointMixin<T extends Constructor<any>>(superClass: T) {
  return class extends superClass {
        /**
         *
         * @type {boolean}
         * @protected
         */
        _stringify = true

        /**
         * Always set Content-Type to application/x-www-form-urlencoded, place in assign last.
         *
         * @returns {AxiosRequestConfig}
         */
        get config(): AxiosRequestConfig {
          return {
            headers: {
              Accept: 'application/json',
              'Content-Type': 'application/x-www-form-urlencoded',
            },
          }
        }

        /**
         * Override to set as version, default to empty
         * e.g. v1, v3, v5 and so on.
         *
         * @returns {string}
         */
        get apiVersion(): string {
          return ''
        }

        /**
         * This it the middle of the url path, usually a group prefix.
         * e.g. api/v1 or user/settings
         *
         * @returns {string}
         */
        get path(): string {
          return urljoin('api', this.apiVersion)
        }

        /**
         *
         * @returns {this}
         */
        preventStringify(): this {
          this._stringify = false
          return this
        }

        /**
         * @returns {boolean}
         */
        get shouldStringify(): boolean {
          return this._stringify
        }

        /**
         *
         * @param {string} url
         * @param {string} method
         * @param {object=} requestConfig
         * @returns {object}
         */
        requestConfig(
          url: string,
          method: Method,
          requestConfig?: AxiosRequestConfig,
        ): AxiosRequestConfig {
          const config: AxiosRequestConfig = super.requestConfig(url, method, requestConfig)
          if (!config.data || typeof config.data !== 'object') {
            return config
          }
          if (this.shouldStringify && !(config.data instanceof FormData)) {
            config.data = qs.stringify(config.data)
          }
          return config
        }
  }
}
