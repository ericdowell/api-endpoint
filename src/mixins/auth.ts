import { AxiosResponse } from 'axios'
import { Constructor } from './types'

// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
export function AuthMixin<T extends Constructor<any>>(superClass: T) {
  return class extends superClass {
    async login<T = any, R = AxiosResponse<T>>(values: {
      email: string
      password: any
      remember?: boolean
    }): Promise<R> {
      const { remember = true } = values
      const data = { remember, email: values.email, password: values.password }
      return this.post('login', { data })
    }

    async logout<T = any, R = AxiosResponse<T>>(): Promise<R> {
      return this.post('logout')
    }

    async register<T = any, R = AxiosResponse<T>>(values: {
      email: string
      emailConfirmation: string
      password: any
      passwordConfirmation: any
      remember?: boolean
      [key: string]: any
    }): Promise<R> {
      const { emailConfirmation, passwordConfirmation, remember = true, ...inputs } = values
      const data = {
        ...inputs,
        // eslint-disable-next-line @typescript-eslint/camelcase
        email_confirmation: emailConfirmation,
        // eslint-disable-next-line @typescript-eslint/camelcase
        password_confirmation: passwordConfirmation,
        remember,
      }
      return this.post('register', { data })
    }

    async requestPasswordReset<T = any, R = AxiosResponse<T>>(email: string): Promise<R> {
      return this.post('password/email', { data: { email } })
    }

    async resetPassword<T = any, R = AxiosResponse<T>>(values: {
      email: string
      token: string
      password: any
      passwordConfirmation: any
    }): Promise<R> {
      const data = {
        email: values.email,
        token: values.token,
        password: values.password,
        // eslint-disable-next-line @typescript-eslint/camelcase
        password_confirmation: values.passwordConfirmation,
      }
      return this.post('password/reset', { data })
    }
  }
}
