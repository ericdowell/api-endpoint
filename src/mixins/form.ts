import { AxiosRequestConfig } from 'axios'
import qs from 'qs'
import { Constructor } from './types'

// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
export function FormMixin<T extends Constructor<any>>(superClass: T) {
  return class extends superClass {
    /** @protected */
    shouldStringify = true

    // Always set Content-Type to application/x-www-form-urlencoded, place in assign last.
    get config(): AxiosRequestConfig {
      return {
        ...super.config,
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded',
          Accept: 'application/json',
          ...super.config.headers,
        },
      }
    }

    preventStringify(): this {
      this.shouldStringify = false
      return this
    }

    requestConfig(requestConfig: AxiosRequestConfig): AxiosRequestConfig {
      const config: AxiosRequestConfig = super.requestConfig(requestConfig)
      if (!config.data || typeof config.data !== 'object') {
        return config
      }
      if (this.shouldStringify && !(config.data instanceof FormData)) {
        config.data = qs.stringify(config.data)
      }
      return config
    }
  }
}
