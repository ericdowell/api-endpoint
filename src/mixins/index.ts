export * from './apiEndpoint'
export * from './handleApi'
export * from './handleError'
export * from './sessionCsrfCookie'
export * from './types'
