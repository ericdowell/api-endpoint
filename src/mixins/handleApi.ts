/* istanbul ignore file */
import { ApiEndpointMixin } from './apiEndpoint'
import { HandleErrorMixin } from './handleError'
import { SessionCsrfCookieMixin } from './sessionCsrfCookie'
import { Constructor } from './types'

// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
export function HandleApiMixin<T extends Constructor<any>> (superClass: T) {
  return class extends SessionCsrfCookieMixin(HandleErrorMixin(ApiEndpointMixin(superClass))) {}
}
