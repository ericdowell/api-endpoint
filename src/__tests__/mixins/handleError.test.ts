import { HandleErrorMixin } from '../../mixins'
import { Endpoint } from '../../endpoint'
import axios from 'axios'
import { BasicMock } from '../mock/axios'

jest.spyOn(axios, 'request').mockImplementation(BasicMock)
// eslint-disable-next-line @typescript-eslint/no-empty-function
jest.spyOn(global.console, 'log').mockImplementation(() => {})

class TestEndpoint extends HandleErrorMixin(Endpoint) {}

describe('HandleErrorMixin', (): void => {
  it.each([
    [{ response: { data: { errors: { key: [] } } } }],
    [undefined],
  ])('handleError returns data key with errors', async(error): Promise<void> => {
    // eslint-disable-next-line jest/prefer-inline-snapshots
    expect(await new TestEndpoint().handleError(error)).toMatchSnapshot()
  })
})
