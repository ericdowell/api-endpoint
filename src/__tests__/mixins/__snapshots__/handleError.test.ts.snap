// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HandleErrorMixin handleError returns data key with errors 1`] = `
Object {
  "config": undefined,
  "data": Object {
    "errors": Object {},
    "message": "An unexpected error has occurred. Please try again.",
  },
  "headers": undefined,
  "status": 500,
  "statusText": "Internal Server Error",
}
`;

exports[`HandleErrorMixin handleError returns data key with errors 2`] = `
Object {
  "config": Object {},
  "data": Object {
    "errors": Object {},
    "message": "An unexpected error has occurred. Please try again.",
  },
  "headers": undefined,
  "status": 500,
  "statusText": "Internal Server Error",
}
`;

exports[`HandleErrorMixin handleError returns data key with errors 3`] = `
Object {
  "data": Object {
    "errors": Object {
      "key": Array [],
    },
  },
}
`;

exports[`HandleErrorMixin handleError returns data key with errors 4`] = `
Object {
  "config": Object {},
  "data": Object {
    "errors": Object {},
    "message": "hello foo bar",
  },
  "headers": Object {},
  "status": 422,
  "statusText": "Unprocessable Entity",
}
`;

exports[`HandleErrorMixin handleError returns data key with errors 5`] = `
Object {
  "config": Object {},
  "data": Object {
    "errors": Object {},
    "message": "An unexpected error has occurred. Please try again.",
  },
  "headers": Object {},
  "status": 400,
  "statusText": "Bad Request",
}
`;
